lib/server/server.mjs

```diff
@@ -0,0 +1,406 @@
+  * to it.
+  *
+  * At least one handler must be present.
+  *
+  * By default, `unknownSubjectHandler` is registered as the default handler.
+  *
+  * @param {...RequestHandler} handlers Default handlers
+  */
+  defaultHandlers (...handlers) {
+    assert(handlers.length > 0, 'At least one default handler must be present!')
+
+    this.#defaultHandlers = handlers
+  }
+
+  /**
+  * @summary Configure the server.
+  *
+  * @description This method expects a method that can be called with the server
+  * as the only argument. This makes it possible to register handler similar to
+  * how `register` can be used in Fastify to add routes.
+  *
+  * @param {function(Server)} configurer Configuration function
+  */
+  configure (configurer) {
+    configurer(this)
+  }
+
+  /**
+  * @summary Connect a stream to the server, listening to its incoming messages.
+  *
+  * @description Internally, the stream will be piped before listening to its
+  * 'data' events. Any errors encountered on either the original or the piped
+  * stream will result in an `error` event on the Server instance.
+  *
+  * After every connected stream, a 'connect' event will be emitted with the
+  * stream as event data.
+  *
+  * In case the stream becomes closed, it will be disconnected automatically,
+  * emitting a 'disconnect' event.
+  *
+  * @param {stream.Duplex} stream Stream
+  */
+  connect (stream) {
+    if (this.#streams.has(stream)) {
+      return
+    }
+
+    const streamContext = new StreamContext({
+      stream,
+      pipe: stream.pipe(ndjson.parse()),
+      handler: data => this.#handleMessage(data, streamContext)
+    })
+
+    this.#logger.debug({ id: streamContext.id, stream }, 'Connecting stream')
+
+    streamContext.pipe.on('data', streamContext.handler)
+    streamContext.stream.on('error', err => this.emit('error',
+      new StreamingError(stream, err))
+    )
+    streamContext.pipe.on('error', err => this.emit('error',
+      new StreamingError(stream, err))
+    )
+
+    stream.on('close', () => {
+      this.#logger.debug({ id: streamContext.id },
+        'Stream closed, disconnecting')
+      this.disconnect(stream)
+    })
+
+    this.#streams.set(stream, streamContext)
+    this.emit('connect', stream)
+  }
+
+  /**
+  * @summary Disconnect the stream, no longer listening to messages coming from
+  * it.
+  *
+  * @description Additionally, a 'disconnect' event is also emitted with the
+  * disconnected stream as event data.
+  *
+  * @param {stream.Duplex} stream Stream
+  */
+  disconnect (stream) {
+    const streamContext = this.#streams.get(stream)
+
+    if (!streamContext) { return }
+
+    this.#logger.debug({ id: streamContext.id, stream }, 'Disconnecting stream')
+    streamContext.pipe.off('data', streamContext.handler)
+    this.#streams.delete(stream)
+    this.emit('disconnect', stream)
+  }
+
+  /**
+  * @param {Message} message
+  * @param {StreamContext} streamContext
+  */
+  #handleMessage (message, streamContext) {
```

Comment:
Quite lenghty method, would it make sense to split it (you actually marked the logically separate sections of it).

Reply:
I was considering to keep it "dumb" i.e. very procedural without any jumps while reading it, but actually the handler-looping is quite terse. Refactored.